update_appearance

TemporalOroboros opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
- Creates `update_name` and `update_desc`
- Creates the wrapper proc `update_appearance` to batch `update_name`, `update_desc`, and `update_icon` together


## Todo list:
- [x] TODO: Finish off the last dozen or so snowflaked `update_icon`s
- [x] TODO: Proofread the entire thing
- [x] TODO: Testing 
  - [X] Unbreak door animations (Moved calling parent to be first so normal door icon updates don't override airlocks updates)
  - [x] Uninvisible airlock control buttons
  - [x] Fix medical kiosks having inverted icon states

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
- Less non-icon handling code in `update_icon` and friends
- Signal hooks for things that want to change names and descriptions
- 99%+ of the changes in this are just from switching everything over to `update_appearance` from `update_icon`

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
code: Adds update_appearance, update_name, and update_icon
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TemporalOroboros committed:
Initial VSCode mass-replace

- Adds and implements [/atom/proc/update_appearance]
- Adds and implements [/atom/proc/update_name]
- Adds and implements [/atom/proc/update_desc]
- Replaces most [/atom/proc/update_icon] references with [/atom/proc/update_appearance]

TemporalOroboros committed:
Cleans up duplicate definitions

- Merges some duplicate definitions into the new procs

TemporalOroboros committed:
Fixes trinary typo

? -> :

MrPerson commented:
How many snowflake update_icon() calls that don't call parent does this leave us with? I know a lot of them were updating name/desc but there were a handful doing other shit.

Unironically in the running for PR of the year 2020, good work

TemporalOroboros committed:
Fixes signal parameters

- Adds a few wrapper procs to eat signal arguments

TemporalOroboros committed:
Straightening out some of the update_icon spaghetti

- Shuffles a lot of the non-icon related dross that was in various update_icon procs to their relevant procs

TemporalOroboros committed:
Merge branch 'master' into update_appearance

LemonInTheDark commented:
![image](https://user-images.githubusercontent.com/58055496/102687083-5cf48180-41a1-11eb-8160-3d7043f5e197.png)

TemporalOroboros committed:
Finishes up the SHOULD_CALL_PARENTing

- Makes all of the base appearance update_x procs SHOULD_CALL_PARENT
- Fixes up some of the update_icon_state procs to not mess around with lighting

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Fixes and misc changes.

- Fixed virology's airlock button being invisible
- Fixed airlocks not doing opening and closing animations
- Fixed the kiosk having inverted icon states for on/off
- Fixed ammo boxes updating custom materials in update_appearance()

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Partially refactors reagent container appearance updates

- Makes reagent container appearances use update_name and update_desc.
- Drinking glasses, shotglasses, and cartons will require their own, dedicated, PR.

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Fixes duplicate proc definition

TemporalOroboros committed:
Fixes compile issue, cleans up a bool

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Cleans up master merge changes

kopoba commented:
emmiters invisible

TemporalOroboros committed:
Fixes invisible emitters

- Adds base icon state

TemporalOroboros commented:
I'm not sure why the tray's aren't harvesting. I'll look into it.

MrStonedOne commented:
Terry experienced instability after this pr was test merged (commit 19e6138)

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master' into update_appearance

TemporalOroboros committed:
Merge remote-tracking branch 'origin/update_appearance' into update_appearance

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Fix merge conflict resolutions

TemporalOroboros committed:
Fix plant trays updating on harvest

LemonInTheDark commented:
How's it going?

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros commented:
@LemonInTheDark Pretty well. I fixed the two things that popped up during the testmerge. I don't know what MSO means by instability but I haven't been able to find anything I changed that would lead to crashes.

LemonInTheDark commented:
@MrStonedOne Could you elaborate?

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge branch 'master' into update_appearance

Rohesie commented on code/game/objects/items/food/pastries.dm:
```diff
@@ -770,15 +770,17 @@
+	name = contents.len ? "stack of pancakes" : initial(name)
+	return ..()
+
+/obj/item/food/pancakes/update_icon(updates=ALL)
+	. = ..(updates & ~UPDATE_OVERLAYS) // Don't update overlays. We're doing that here
```
```suggestion
/obj/item/food/pancakes/update_icon(updates = (ALL & ~UPDATE_OVERLAYS)) // Don't update overlays. We're doing that here
	. = ..()
```
or
```suggestion
/obj/item/food/pancakes/update_icon(updates = (ALL))
	updates &= ~UPDATE_OVERLAYS // Don't update overlays. We're doing that here
	. = ..()
```

TemporalOroboros replied:
It should be done in the next push.

Rohesie commented:
Massive PR, could use atomization (lots of code cleaning), but it's a very good change.
Fix the conflict, test it locally, and we can get it TMed again.

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Fix master merge issues

- Reimplements the marker beacon icon prefix
- Changes a couple update_icon calls to update_appearance calls
- Removes an extra tab from mechcode.

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Fixes merge errors and uncopypasta's pipe layers.

TemporalOroboros committed:
Touches up morgue code.

TemporalOroboros committed:
Ho, linter! Shut thee up.

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Removes an extraneous tab

Rohesie commented on code/__DEFINES/dcs/signals.dm:
```diff
@@ -74,13 +74,30 @@
+///from base of [/atom/proc/update_appearance]: (updates)
+#define COMSIG_ATOM_UPDATE_APPEARANCE	"atom_update_appearance"
+	/// If returned from [COMSIG_ATOM_UPDATE_APPEARANCE] it prevents the atom from updating its name.
+	#define COMSIG_ATOM_NO_UPDATE_NAME			UPDATE_NAME
+	/// If returned from [COMSIG_ATOM_UPDATE_APPEARANCE] it prevents the atom from updating its desc.
+	#define COMSIG_ATOM_NO_UPDATE_DESC			UPDATE_DESC
+	/// If returned from [COMSIG_ATOM_UPDATE_APPEARANCE] it prevents the atom from updating its icon.
+	#define COMSIG_ATOM_NO_UPDATE_ICON			UPDATE_ICON
+///from base of [/atom/proc/update_name]: (updates)
+#define COMSIG_ATOM_UPDATE_NAME			"atom_update_name"
+///from base of [/atom/proc/update_desc]: (updates)
+#define COMSIG_ATOM_UPDATE_DESC			"atom_update_desc"
+///from base of [/atom/update_icon]: ()
+#define COMSIG_ATOM_UPDATE_ICON			"atom_update_icon"
+	/// If returned from [COMSIG_ATOM_UPDATE_ICON] it prevents the atom from updating its icon state.
+	#define COMSIG_ATOM_NO_UPDATE_ICON_STATE	UPDATE_ICON_STATE
+	/// If returned from [COMSIG_ATOM_UPDATE_ICON] it prevents the atom from updating its overlays.
+	#define COMSIG_ATOM_NO_UPDATE_OVERLAYS		UPDATE_OVERLAYS
+///from base of [atom/update_icon_state]: ()
+#define COMSIG_ATOM_UPDATE_ICON_STATE	"atom_update_icon_state"
+///from base of [/atom/update_overlays]: (list/new_overlays)
+#define COMSIG_ATOM_UPDATE_OVERLAYS		"atom_update_overlays"
```
Could you remove the extra spaces, and use the standard in the file?

TemporalOroboros replied:
Done

Rohesie commented on code/game/atoms.dm:
```diff
@@ -640,31 +640,64 @@
 /// Updates the icon state of the atom
 /atom/proc/update_icon_state()
+	SHOULD_CALL_PARENT(TRUE)
+	. = SEND_SIGNAL(src, COMSIG_ATOM_UPDATE_ICON_STATE)
```
```suggestion
	return SEND_SIGNAL(src, COMSIG_ATOM_UPDATE_ICON_STATE)
```

TemporalOroboros replied:
Done

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Commits @Rohesie's Suggestions

- Removes an extraneous . assignment
- Removes extra signal define whitespace

Rohesie commented:
Is this locally tested and working?
Should it get a TM?

TemporalOroboros commented:
@Rohesie I have tested this on local and it appears to be working. A TM would be appriciated unless a recent merge conflicts with it.

Rohesie commented:
> 
> 
> @Rohesie I have tested this on local and it appears to be working. A TM would be appriciated unless a recent merge conflicts with it.

Will need a conflict merge and new local test, even if a quick one.

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Fix null access in drinking glasses

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Fixes incorrect var name from master merge

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Cleans up some master merge changes

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge branch 'master' into update_appearance

Rohesie commented:
Confirm you've tested it locally before unmarking it from draft.

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Fixes shot drinking glasses and the combo screen object.

TemporalOroboros commented:
Did a bit of local testing. Nothing appeared to be broken.

Cyberboss reviewed (changes requested):
Fix conflicts and lets get it tested then

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge master fixes and misc updates

- Moves a whole bunch of parent calls in update_icon_state to the end of the proc to give the signal more room to work with.
- Adds a whole bunch of early returns to various update_icon_state and update_overlays procs.
- Refactors APC icon update code. (TODO: Refactor APCs)

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros commented:
Did some local testing. Everything seemed to be working fine.

AnturK commented:
Conflicts.

Also i hate to say this so late but update_appearances seems like kinda misleading name if it's about more than what appearance involves.

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros commented:
I consider the name and description to be part of the appearance.

TemporalOroboros commented:
Seems to be working.

TemporalOroboros committed:
Merge branch 'master' into update_appearance

Timberpoes commented on code/game/machinery/recharger.dm:
```diff
@@ -163,24 +163,29 @@
-		if(using_power)
-			SSvis_overlays.add_vis_overlay(src, icon, "recharger-charging", layer, plane, dir, alpha)
-			SSvis_overlays.add_vis_overlay(src, icon, "recharger-charging", EMISSIVE_LAYER, EMISSIVE_PLANE, dir, alpha)
-		else
-			SSvis_overlays.add_vis_overlay(src, icon, "recharger-full", layer, plane, dir, alpha)
-			SSvis_overlays.add_vis_overlay(src, icon, "recharger-full", EMISSIVE_LAYER, EMISSIVE_PLANE, dir, alpha)
-	else
+	if(!charging)
 		SSvis_overlays.add_vis_overlay(src, icon, "recharger-empty", layer, plane, dir, alpha)
 		SSvis_overlays.add_vis_overlay(src, icon, "recharger-empty", EMISSIVE_LAYER, EMISSIVE_PLANE, dir, alpha)
+
+	if(using_power)
+		SSvis_overlays.add_vis_overlay(src, icon, "recharger-charging", layer, plane, dir, alpha)
+		SSvis_overlays.add_vis_overlay(src, icon, "recharger-charging", EMISSIVE_LAYER, EMISSIVE_PLANE, dir, alpha)
+		return
+
+	SSvis_overlays.add_vis_overlay(src, icon, "recharger-full", layer, plane, dir, alpha)
+	SSvis_overlays.add_vis_overlay(src, icon, "recharger-full", EMISSIVE_LAYER, EMISSIVE_PLANE, dir, alpha)
```
You need to early return here, otherwise you'll fall through to apply to the recharger-full overlay when the charger is empty, and this overlay will sit ontop of it.

Timberpoes commented on code/game/machinery/recharger.dm:
```diff
@@ -163,24 +163,29 @@
-			SSvis_overlays.add_vis_overlay(src, icon, "recharger-full", layer, plane, dir, alpha)
-			SSvis_overlays.add_vis_overlay(src, icon, "recharger-full", EMISSIVE_LAYER, EMISSIVE_PLANE, dir, alpha)
-	else
+	if(!charging)
 		SSvis_overlays.add_vis_overlay(src, icon, "recharger-empty", layer, plane, dir, alpha)
 		SSvis_overlays.add_vis_overlay(src, icon, "recharger-empty", EMISSIVE_LAYER, EMISSIVE_PLANE, dir, alpha)
+
+	if(using_power)
```
```suggestion
	if(!charging)
		SSvis_overlays.add_vis_overlay(src, icon, "recharger-empty", layer, plane, dir, alpha)
		SSvis_overlays.add_vis_overlay(src, icon, "recharger-empty", EMISSIVE_LAYER, EMISSIVE_PLANE, dir, alpha)
		return
	if(using_power)
```

Timberpoes commented on code/game/machinery/recharger.dm:
```diff
@@ -163,24 +163,29 @@
+
 /obj/machinery/recharger/update_overlays()
 	. = ..()
-	SSvis_overlays.remove_vis_overlay(src, managed_vis_overlays)
```
Make sure that overlays are being properly removed by update_overlay when this proc is called. When I admin proc called update_overlays on the recharger to diagnose the other issue, it duplicated the overlays. This is probably a bad thing.

![image](https://user-images.githubusercontent.com/24975989/107249685-7e4c4e00-6a2b-11eb-893b-6a56e8ff6593.png)

TemporalOroboros replied:
@Timberpoes As far as I can tell it's always been assumed that update_overlays would not be directly called. It never removed the normal overlays either.

TemporalOroboros replied:
See the definition for the base of update_icon on master.

TemporalOroboros replied:
Though, in this case I think I can move the vis overlay removal into the proc base.

TemporalOroboros committed:
Update code/game/machinery/recharger.dm

Co-authored-by: Timberpoes <silent_insomnia_pp@hotmail.co.uk>

Yenwodyah commented:
ORM is invisible and rechargers are always showing their charging icon

Timberpoes commented:
![Ad4pZEMOMY](https://user-images.githubusercontent.com/24975989/107272232-4b16b880-6a45-11eb-9b80-03f3bc57770d.gif)

Closets have no open state.

Timberpoes commented:
Grilles are getting null icon states : )

![image](https://user-images.githubusercontent.com/24975989/107272388-874a1900-6a45-11eb-9061-e976a4ca4159.png)

Timberpoes reviewed (changes requested):
See above reports. I believe they're all linked to this PR. Please investigate and fixies.

TemporalOroboros committed:
Various fixes

- Fixes grilles showing up as :) tables.
- Fixes the ORM turning invisible.
- Fixes lockers always appearing to be closed.
- Fixes flashes getting stuck with an eternal flashing overlay.
- Makes the recharger use its base icon state.

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros commented:
I fixed everything you pointed out and also flashes. This should be ready to TM again.

Cyberboss commented:
Beatings will continue until morale improves.

TemporalOroboros committed:
Merge branch 'master' into update_appearance

Cyberboss commented:
Confrirtuirtuerituksdljfklsdjflksdf

Qustinnus commented:
i dont see any merge conflicts

Cyberboss commented:
You can't see them in draft mode.

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Fixes nukeops tablets.

- Fixes nuclear operator tablets winding up with an invalid icon state
- Makes some plant genes use update_appearance rather than update_icon.

TemporalOroboros commented:
@Cyberboss It is ready to TM again.

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Merge branch 'master' into update_appearance

NEDM5333 commented:
Bone axes turn to normal axes when dropped.
Mining vendors invisible

TemporalOroboros committed:
Merge branch 'master' into update_appearance

Timberpoes commented:
Reports of invisible blobs too - This was due to the blob being qdelled but hard deleting, so the player still had control of the blob's AI camera for awhile. The AI camera also qdelled. A result of this TM?

TemporalOroboros committed:
Fixes cannonballs

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Fixes tabs after code

TemporalOroboros commented:
@NEDM5333 Both should be fixed in the next push.
@Timberpoes Got it. The blob reworked repathed the core and I didn't notice when I merged master. It should be fixed in the next push.

TemporalOroboros committed:
Master merge and dual wielding fixes + removing a bunch of update_icon calls

TemporalOroboros committed:
Merge branch 'master' into update_appearance

MrMelbert commented:
testmerged on #156451, hydroponics trays never had their nutrient light shut off even after refilling them

TemporalOroboros commented:
I"'ll look into it.

NEDM5333 commented:
Don't know if you've already fixed this, but during the testmerge yesterday I noticed that all positronic brains had the same sprite as empty MMIs, no matter their state

MrMelbert commented:
![image](https://user-images.githubusercontent.com/51863163/108036248-9872c700-6ffd-11eb-99db-33870d9b5325.png)
syndie gun turrets are normal turret sprites

Cyberboss commented:
Conflicts again.

I want to deeply thank you for this PR and the effort you're putting into it. This is a much needed refactor.

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Miscellaneous fixes

- Fixes syndicate turrets having covers
- Fixes cigarette boxes turning invisible
- Fixes hydroponics tray lights
- Fixes posibrains turning into MMIs

TemporalOroboros committed:
Merge branch 'master' into update_appearance

Cyberboss commented:
The conflicts, they don't stop 😭

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Fixes IV drip

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros commented:
@Cyberboss Conflixed

Cyberboss commented:
@tgstation/commit-access please try and get this PR priority test merged before merging other things. It's conflicting faster than I can get to it.

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Fixes ammo

TemporalOroboros committed:
Merge branch 'master' into update_appearance

TemporalOroboros committed:
Removes projectile files from the wrong branch

TemporalOroboros commented:
@Cyberboss Conflicts fixed

Rohesie commented:
This PR has been up since last year. We'll have to test it live and fix the issues that come up, or the conflicts will never end.
